File: commands/src/servo_comm.cpp
#include "servo_comm.hpp"

// Destructor
ServoComm::~ServoComm() {

    // if the serial port is open, close it
    if (this->teensy_serial_port >= 0){
        close(this->teensy_serial_port);
    }

    std::cout << "Closed Teensy serial port." << std::endl;
};

// intialize the serial communication
bool ServoComm::initialize_comm(const std::string& port) {

    // open Teensy serial port
    this->teensy_serial_port = open(port.c_str(), O_RDWR);

    // serial port failed to open
    if (this->teensy_serial_port < 0) {

        // print error message
        std::cerr << "Failed to open Teensy serial port at " << port << "." << std::endl;
        return false;
    }

    // serial port opened successfully
    else {

        // setup the serial port settings
        this->setup_serial_port();

        // print success message
        std::cout << "Opened Teensy serial port at " << port << "." << std::endl;
        return true;
    }
}

// setup the serial port settings
void ServoComm::setup_serial_port() {

    // configure serial port settings
    tcgetattr(this->teensy_serial_port, &this->tty);

    // set baud rate
    cfsetispeed(&this->tty, B9600);
    cfsetospeed(&this->tty, B9600);

    // set other serial port settings
    this->tty.c_cflag &= ~PARENB; // no parity
    this->tty.c_cflag &= ~CSTOPB; // one stop bit
    this->tty.c_cflag &= ~CSIZE;
    this->tty.c_cflag |= CS8;     // 8 bits per byte
    this->tty.c_cflag |= CREAD | CLOCAL;

    // apply settings
    tcsetattr(this->teensy_serial_port, TCSANOW, &this->tty);
}

// command the servo to be in the open positions
void ServoComm::command_opened_pos() {

    // send the open command to the Teensy
    std::string command;
    command = "0";

    // write to the serial port
    write(this->teensy_serial_port, command.c_str(), 1);
}

// command the servo to be in the closed positions
void ServoComm::command_locked_pos() {

    // send the closed command to the Teensy
    std::string command;
    command = "1";

    // write to the serial port
    write(this->teensy_serial_port, command.c_str(), 1);
}

File: src/servo_comm.py
#
#  Python servo controller class (mirrors src/servo_comm.hpp / src/servo_comm.cpp).
#  Communicates with the Teensy over a serial port.
#

import serial


# servo controller class
class ServoComm:

    # constructor
    def __init__(self):
        self.teensy_serial_port = None

    # destructor: close the serial port if it is open
    def __del__(self):
        if self.teensy_serial_port is not None and self.teensy_serial_port.is_open:
            self.teensy_serial_port.close()
            print("Closed Teensy serial port.")

    # initialize and setup serial communication
    def initialize_comm(self, port):

        try:
            # open Teensy serial port with 9600 baud, 8N1
            self.teensy_serial_port = serial.Serial(
                port=port,
                baudrate=9600,
                bytesize=serial.EIGHTBITS,
                parity=serial.PARITY_NONE,
                stopbits=serial.STOPBITS_ONE,
                timeout=1,
            )

        # serial port failed to open
        except serial.SerialException:
            print(f"Failed to open Teensy serial port at {port}.")
            return False

        # serial port opened successfully
        print(f"Opened Teensy serial port at {port}.")
        return True

    # command the servo to be in the open positions
    def command_opened_pos(self):
        self.teensy_serial_port.write(b"0")

    # command the servo to be in the closed positions
    def command_locked_pos(self):
        self.teensy_serial_port.write(b"1")
